# Lista 4, Métodos Numéricos 2 
Javier Alejandro Ovalle Chiquín, 22103  
Ricardo Josué Morales Contreras, 22289

#### Problema 1. Modelo de Prooducción, Periodo único

Definimos las variables de decisión:

- $x_1$: Número de abrigos de piel con capucha producidos.
- $x_2$: Número de chamarras con relleno de plumas de ganso producidas.
- $x_3$: Número de pantalones con aislamiento producidos.
- $x_4$: Número de guantes producidos.
- $s_1, s_2, s_3, s_4 \geq 0$: Variables de faltante (shortage) para cada producto, respectivamente.

Las demandas firmes son: $d_1 = 800$, $d_2 = 750$, $d_3 = 600$, $d_4 = 500$.

Las utilidades por unidad son: 30 para abrigos, 40 para chamarras, 20 para pantalones, 10 para guantes.

Las penalizaciones por unidad no surtida son: 15 para abrigos, 20 para chamarras, 10 para pantalones, 8 para guantes.

**Función objetivo** (maximizar utilidad neta):  
$$
\max z = 30x_1 + 40x_2 + 20x_3 + 10x_4 - 15s_1 - 20s_2 - 10s_3 - 8s_4
$$

**Restricciones de demanda** (se debe cubrir la demanda o incurrir en faltante):  
$$
x_1 + s_1 = 800
$$  
$$
x_2 + s_2 = 750
$$  
$$
x_3 + s_3 = 600
$$  
$$
x_4 + s_4 = 500
$$

**Restricciones de capacidad por departamento** (en horas):  
- Corte: $0.30x_1 + 0.30x_2 + 0.25x_3 + 0.15x_4 \leq 1000$  
- Aislamiento: $0.25x_1 + 0.35x_2 + 0.30x_3 + 0.10x_4 \leq 1000$  
- Costura: $0.45x_1 + 0.50x_2 + 0.40x_3 + 0.22x_4 \leq 1000$  
- Empaque: $0.15x_1 + 0.15x_2 + 0.10x_3 + 0.05x_4 \leq 1000$

**Restricciones de no negatividad**:  
$$
x_1, x_2, x_3, x_4, s_1, s_2, s_3, s_4 \geq 0
$$


Solución del modelo con la librería pulp sin restricción de enteros.

In [2]:
import pulp

# Crear el problema
prob = pulp.LpProblem("Produccion_Optima", pulp.LpMaximize)

# Variables continuas
x1 = pulp.LpVariable("Abrigos", lowBound=0)
x2 = pulp.LpVariable("Chamarras", lowBound=0)
x3 = pulp.LpVariable("Pantalones", lowBound=0)
x4 = pulp.LpVariable("Guantes", lowBound=0)
s1 = pulp.LpVariable("Faltante_Abrigos", lowBound=0)
s2 = pulp.LpVariable("Faltante_Chamarras", lowBound=0)
s3 = pulp.LpVariable("Faltante_Pantalones", lowBound=0)
s4 = pulp.LpVariable("Faltante_Guantes", lowBound=0)

# Objetivo
prob += 30*x1 + 40*x2 + 20*x3 + 10*x4 - 15*s1 - 20*s2 - 10*s3 - 8*s4

# Restricciones de demanda
prob += x1 + s1 == 800
prob += x2 + s2 == 750
prob += x3 + s3 == 600
prob += x4 + s4 == 500

# Restricciones de departamentos
prob += 0.30*x1 + 0.30*x2 + 0.25*x3 + 0.15*x4 <= 1000, "Corte"
prob += 0.25*x1 + 0.35*x2 + 0.30*x3 + 0.10*x4 <= 1000, "Aislamiento"
prob += 0.45*x1 + 0.50*x2 + 0.40*x3 + 0.22*x4 <= 1000, "Costura"
prob += 0.15*x1 + 0.15*x2 + 0.10*x3 + 0.05*x4 <= 1000, "Empaque"

# Resolver
prob.solve(pulp.PULP_CBC_CMD(msg=0))

# Resultados
print("Valor objetivo:", pulp.value(prob.objective))
print("x1:", x1.varValue)
print("x2:", x2.varValue)
print("x3:", x3.varValue)
print("x4:", x4.varValue)
print("s1:", s1.varValue)
print("s2:", s2.varValue)
print("s3:", s3.varValue)
print("s4:", s4.varValue)

Valor objetivo: 64625.0
x1: 800.0
x2: 750.0
x3: 387.5
x4: 500.0
s1: 0.0
s2: 0.0
s3: 212.5
s4: 0.0


Estos resultados coinciden exáctamente con los mostrados por simplex de excel

Solución del problema, añadiendo la restricción de que las variables sean enteras:

In [3]:
import pulp

# Crear el problema
prob = pulp.LpProblem("Produccion_Entera", pulp.LpMaximize)

# Variables enteras
x1 = pulp.LpVariable("Abrigos", lowBound=0, cat='Integer')
x2 = pulp.LpVariable("Chamarras", lowBound=0, cat='Integer')
x3 = pulp.LpVariable("Pantalones", lowBound=0, cat='Integer')
x4 = pulp.LpVariable("Guantes", lowBound=0, cat='Integer')
s1 = pulp.LpVariable("Faltante_Abrigos", lowBound=0, cat='Integer')
s2 = pulp.LpVariable("Faltante_Chamarras", lowBound=0, cat='Integer')
s3 = pulp.LpVariable("Faltante_Pantalones", lowBound=0, cat='Integer')
s4 = pulp.LpVariable("Faltante_Guantes", lowBound=0, cat='Integer')

# Objetivo 
prob += 30*x1 + 40*x2 + 20*x3 + 10*x4 - 15*s1 - 20*s2 - 10*s3 - 8*s4

# Restricciones 
prob += x1 + s1 == 800
prob += x2 + s2 == 750
prob += x3 + s3 == 600
prob += x4 + s4 == 500
prob += 0.30*x1 + 0.30*x2 + 0.25*x3 + 0.15*x4 <= 1000, "Corte"
prob += 0.25*x1 + 0.35*x2 + 0.30*x3 + 0.10*x4 <= 1000, "Aislamiento"
prob += 0.45*x1 + 0.50*x2 + 0.40*x3 + 0.22*x4 <= 1000, "Costura"
prob += 0.15*x1 + 0.15*x2 + 0.10*x3 + 0.05*x4 <= 1000, "Empaque"

# Resolver
prob.solve(pulp.PULP_CBC_CMD(msg=0))

# Resultados
print("Valor objetivo:", pulp.value(prob.objective))
print("x1:", x1.varValue)
print("x2:", x2.varValue)
print("x3:", x3.varValue)
print("x4:", x4.varValue)
print("s1:", s1.varValue)
print("s2:", s2.varValue)
print("s3:", s3.varValue)
print("s4:", s4.varValue)

Valor objetivo: 64622.0
x1: 800.0
x2: 750.0
x3: 388.0
x4: 499.0
s1: 0.0
s2: 0.0
s3: 212.0
s4: 1.0


Como se puede observar varian los resultados en $x_3$ que son el número de pantalones producidos, ya que era la solución que no quedaba entera, modificandolo a que sean 388 pantalones y restanto 1 al número de guantes producidos teniendo $212$ pantalones faltantes y 1 guante faltante llegando a la solución optima de $64622$. 

Lo interesante viene al compararlo con el método simplex de excel, ya que la solución optima fue dejando $387$ pantalones y $213$ faltantes, llegando a una subsolucion de $64610$, la cual es menos a la solución propuesta por pulp. Esto probablemente se debe a las limitantes de las interaciones con excel.

#### Problema 2. Modelo de Producción, periodos múltiples 

Definimos los índices: $t = 1, 2, \dots, 6$ para los meses.

**Datos:**  
- Demanda $d_t$: 180, 250, 190, 140, 220 y 250 ventanas.  
- Costo de producción por ventana $c_t$: \$50, \$45, \$55, \$52, \$48 y \$50 durante los próximos seis meses.  
- Costo de almacenamiento por ventana a fin de mes $h_t$: \$8 por ventana por mes en los meses 1, 5 y 6. En los meses 2, 3 y 4, los costos por almacenaje son de \$10 por ventana.  
- Capacidad máxima de producción: 225 ventanas cada mes.  
- Inventario inicial $I_0 = 0$, y asumimos $I_6 = 0$ (sin inventario final para evitar costos innecesarios).

**Variables de decisión:**  
- $P_t \geq 0$: Producción en el mes $t$ (ventanas).  
- $I_t \geq 0$: Inventario al final del mes $t$ (ventanas).

**Función objetivo** (minimizar costo total: producción + almacenamiento):  
$$
\min z = \sum_{t=1}^6 \left( c_t P_t + h_t I_t \right)
$$

En la sumatoria se estan considerando 4 variables, sin embargo 2 de ellas serán solo valores numéricos por cada mes, siendo estos el costo  de producción por ventana y el costo de almacenaje por mes

**Restricciones:**  
1. **Balance de inventario** (producción + inventario inicial = demanda + inventario final):  
$$
I_{t-1} + P_t = d_t + I_t, \quad \forall t = 1, \dots, 6
$$  
(con $I_0 = 0$).

2. **Capacidad de producción**:  
$$
P_t \leq 225, \quad \forall t = 1, \dots, 6
$$

3. **No negatividad**:  
$$
P_t \geq 0, \quad I_t \geq 0, \quad \forall t
$$

In [4]:
import pulp

# Datos
demandas = [180, 250, 190, 140, 220, 250]
costos_prod = [50, 45, 55, 52, 48, 50]
costos_inv = [8, 10, 10, 10, 8, 8]
capacidad = 225
meses = range(1,7)

# Problema continuo
prob_cont = pulp.LpProblem("Produccion_Multiperiodo_Cont", pulp.LpMinimize)

P = pulp.LpVariable.dicts("Produccion", meses, lowBound=0, upBound=capacidad)
I0 = pulp.LpVariable("I0", lowBound=0, upBound=0)
I = pulp.LpVariable.dicts("Inventario", meses, lowBound=0)
I6 = pulp.LpVariable("I6", lowBound=0)

# Objetivo
prob_cont += pulp.lpSum(costos_prod[t-1]*P[t] + costos_inv[t-1]*I[t] for t in meses)

# Balance
prev_I = I0
for t in meses:
    if t < 6:
        prob_cont += prev_I + P[t] == demandas[t-1] + I[t]
        prev_I = I[t]
    else:
        prob_cont += prev_I + P[t] == demandas[t-1] + I6
prob_cont += I6 == 0

prob_cont.solve(pulp.PULP_CBC_CMD(msg=0))

# Resultados
print("Objetivo:", pulp.value(prob_cont.objective))
for t in meses:
    print(f"Mes {t}: P={P[t].varValue:.2f}, I_end={I[t].varValue:.2f}")

Objetivo: 61795.0
Mes 1: P=205.00, I_end=25.00
Mes 2: P=225.00, I_end=0.00
Mes 3: P=190.00, I_end=0.00
Mes 4: P=160.00, I_end=20.00
Mes 5: P=225.00, I_end=25.00
Mes 6: P=225.00, I_end=0.00


La estrategia óptima produce al máximo (225) en meses baratos (2 y 5, con $ c_2=45 $, $ c_5=48 $) para cubrir parte de las demandas altas en 2 y 6, almacenando el excedente. En mes 1, produce extra (205 > 180) para ayudar en mes 2. En mes 4, produce 160 > 140 pero almacena para mes 5 y 6. Mes 3 produce exacto ya que $ c_3=55 $ es alto.

Para la comparativa con la solución no óptima, que es producir exactamente la cantidad de la demanda por cada mes se observa a simple vista que no se llegará a cumplir la demanda de los meses 2 y 6 ya que la capacidad máxima de producir es de 225 mientras que la demanda será de 250 en ambos meses, sin embargo se obtiene una solución la cual sería calculada por $180 \cdot 50+ 225 \cdot 45 + 190 \cdot 55 + 140 \cdot 52 + 220 \cdot 48 + 225 \cdot 50 = 58,665 $ sin llegar a la demanda por la falta de capacidad de producción.


#### Problema 3. Modelo de renovaión urbana

Definimos las variables de decisión (número de unidades a construir, $\geq 0$):

- $x_1$: Número de unidades sencillas (unifamiliares).
- $x_2$: Número de unidades dobles.
- $x_3$: Número de unidades triples.
- $x_4$: Número de unidades cuádruples.

**Datos clave:**
- Terreno disponible de demolición: Máximo 300 casas $\times$ 0.25 acres/casa = 75 acres.
- Área usable para construcción: $75 \times (1 - 0.15) = 63.75$ acres (15% para calles, espacios y servicios).
- Tamaños de lotes (acres/unidad): 0.18, 0.28, 0.40, 0.50.
- Ingresos por impuestos/unidad: \$1000, \$1900, \$2700, \$3400.
- Costos de construcción/unidad: \$50,000, \$70,000, \$130,000, \$160,000.
- Presupuesto máximo: \$15,000,000.
- (El costo de demolición de \$2000/casa es fijo y no afecta el objetivo, ya que se asume demolición según necesidad hasta 300 casas).

**Función objetivo** (maximizar recaudación de impuestos):  
$$
\max z = 1000 x_1 + 1900 x_2 + 2700 x_3 + 3400 x_4
$$

**Restricciones:**  
1. **Disponibilidad de terreno**:  
$$
0.18 x_1 + 0.28 x_2 + 0.40 x_3 + 0.50 x_4 \leq 63.75
$$

2. **Unidades triples y cuádruples $\geq$ 25\% del total** (de $U = x_1 + x_2 + x_3 + x_4$):  
$$
x_3 + x_4 \geq 0.25 U \implies -x_1 - x_2 + 3 x_3 + 3 x_4 \geq 0
$$

3. **Unidades sencillas $\geq$ 20\% del total**:  
$$
x_1 \geq 0.20 U \implies 0.80 x_1 - 0.20 x_2 - 0.20 x_3 - 0.20 x_4 \geq 0
$$

4. **Unidades dobles $\geq$ 10\% del total**:  
$$
x_2 \geq 0.10 U \implies -0.10 x_1 + 0.90 x_2 - 0.10 x_3 - 0.10 x_4 \geq 0
$$

5. **Presupuesto de construcción**:  
$$
50{,}000 x_1 + 70{,}000 x_2 + 130{,}000 x_3 + 160{,}000 x_4 \leq 15{,}000{,}000
$$

6. **No negatividad**:  
$$
x_1, x_2, x_3, x_4 \geq 0
$$

In [6]:
import pulp

# Problema continuo
prob_cont = pulp.LpProblem("Renovacion_Urbana_Cont", pulp.LpMaximize)

# Variables
x1 = pulp.LpVariable("Sencillas", lowBound=0)
x2 = pulp.LpVariable("Dobles", lowBound=0)
x3 = pulp.LpVariable("Triples", lowBound=0)
x4 = pulp.LpVariable("Cuadruples", lowBound=0)

# Objetivo: maximizar impuestos
prob_cont += 1000*x1 + 1900*x2 + 2700*x3 + 3400*x4

# Restriccion tierra usable: 75 acres * 0.85 = 63.75
prob_cont += 0.18*x1 + 0.28*x2 + 0.40*x3 + 0.50*x4 <= 63.75, "Tierra"

# Triples y cuadruples >=25% total
prob_cont += 3*x3 + 3*x4 - x1 - x2 >= 0, "Triples_Cuadruples_25"

# Sencillas >=20%
prob_cont += 0.8*x1 - 0.2*x2 - 0.2*x3 - 0.2*x4 >= 0, "Sencillas_20"

# Dobles >=10%
prob_cont += -0.1*x1 + 0.9*x2 - 0.1*x3 - 0.1*x4 >= 0, "Dobles_10"

# Presupuesto construccion <=15M
prob_cont += 50000*x1 + 70000*x2 + 130000*x3 + 160000*x4 <= 15000000, "Presupuesto"

# Resolver
prob_cont.solve(pulp.PULP_CBC_CMD(msg=0))

# Resultados
print("Objetivo continuo:", pulp.value(prob_cont.objective))
print("x1:", x1.varValue)
print("x2:", x2.varValue)
print("x3:", x3.varValue)
print("x4:", x4.varValue)

Objetivo continuo: 355555.5512
x1: 37.037037
x2: 101.85185
x3: 46.296296
x4: 0.0


Se llegó a la solución optima donde se cumple que las restricciones, dando como resultado $37.037$ de unidades sencillas, $101.85$ unidades dobles, $46.0296$ unidades triples y ninguna cuadruple.

In [7]:
import pulp

# Problema entero
prob_ent = pulp.LpProblem("Renovacion_Urbana_Ent", pulp.LpMaximize)

# Variables enteras
x1 = pulp.LpVariable("Sencillas", lowBound=0, cat='Integer')
x2 = pulp.LpVariable("Dobles", lowBound=0, cat='Integer')
x3 = pulp.LpVariable("Triples", lowBound=0, cat='Integer')
x4 = pulp.LpVariable("Cuadruples", lowBound=0, cat='Integer')

# Objetivo: maximizar impuestos
prob_ent += 1000*x1 + 1900*x2 + 2700*x3 + 3400*x4

# Restriccion tierra usable: 75 acres * 0.85 = 63.75
prob_ent += 0.18*x1 + 0.28*x2 + 0.40*x3 + 0.50*x4 <= 63.75, "Tierra"

# Triples y cuadruples >=25% total
prob_ent += 3*x3 + 3*x4 - x1 - x2 >= 0, "Triples_Cuadruples_25"

# Sencillas >=20%
prob_ent += 0.8*x1 - 0.2*x2 - 0.2*x3 - 0.2*x4 >= 0, "Sencillas_20"

# Dobles >=10%
prob_ent += -0.1*x1 + 0.9*x2 - 0.1*x3 - 0.1*x4 >= 0, "Dobles_10"

# Presupuesto construccion <=15M
prob_ent += 50000*x1 + 70000*x2 + 130000*x3 + 160000*x4 <= 15000000, "Presupuesto"

# Resolver
prob_ent.solve(pulp.PULP_CBC_CMD(msg=0))

# Resultados
print("Objetivo entero:", pulp.value(prob_ent.objective))
print("x1:", x1.varValue)
print("x2:", x2.varValue)
print("x3:", x3.varValue)
print("x4:", x4.varValue)

Objetivo entero: 355400.0
x1: 36.0
x2: 99.0
x3: 31.0
x4: 14.0


Nótese que la solución entera varia conforme la anterior, ahora incluyendo 14 viviendas cuadruples las cuales no se usaban en la solución continua, en este caso se utiliza el presupuesto en su totalidad, llegando a los 15 millones de dólares. 

#### Problema 4. Modelo de asignación de horarios

### 4. (Modelo de asignación de horarios)


Definimos las variables de decisión (cantidad de autobuses que inician en cada turno, $\geq 0$):

- $x_1$: Número de autobuses que inician en el turno 1 
- $x_2$: Número de autobuses que inician en el turno 2 
- $x_3$: Número de autobuses que inician en el turno 3 
- $x_4$: Número de autobuses que inician en el turno 4 
- $x_5$: Número de autobuses que inician en el turno 5 
- $x_6$: Número de autobuses que inician en el turno 6 

Cada autobús opera exactamente 8 horas continuas (cubriendo dos periodos consecutivos de 4 horas). Los periodos de demanda (y requerimientos mínimos de autobuses activos) son:

- Periodo 1: 4 autobuses.
- Periodo 2: 8 autobuses.
- Periodo 3: 10 autobuses.
- Periodo 4: 7 autobuses.
- Periodo 5: 12 autobuses.
- Periodo 6: 4 autobuses.

**Función objetivo** (minimizar el número total de autobuses):  
$$
\min z = x_1 + x_2 + x_3 + x_4 + x_5 + x_6
$$

**Restricciones** (cobertura de demanda en cada periodo):  
1. **Periodo 1**: $x_1 + x_6 \geq 4$  
2. **Periodo 2**: $x_1 + x_2 \geq 8$  
3. **Periodo 3**: $x_2 + x_3 \geq 10$  
4. **Periodo 4**: $x_3 + x_4 \geq 7$  
5. **Periodo 5**: $x_4 + x_5 \geq 12$  
6. **Periodo 6**: $x_5 + x_6 \geq 4$  

**No negatividad**:  
$$
x_1, x_2, x_3, x_4, x_5, x_6 \geq 0
$$

Este es un problema de programación lineal con 6 variables y 6 restricciones.

In [8]:
import pulp

# Problema
prob = pulp.LpProblem("Asignacion_Horarios_Buses", pulp.LpMinimize)

# Variables continuas
x = pulp.LpVariable.dicts("x", range(1,7), lowBound=0)

# Objetivo: minimizar total buses
prob += pulp.lpSum(x[i] for i in range(1,7))

# Demandas por periodo (4h)
demandas = [4,8,10,7,12,4]

# Restricciones por periodo
# Periodo 1 (00-04): x1 + x6 >= 4
prob += x[1] + x[6] >= demandas[0], "P1"

# P2 (04-08): x1 + x2 >=8
prob += x[1] + x[2] >= demandas[1], "P2"

# P3 (08-12): x2 + x3 >=10
prob += x[2] + x[3] >= demandas[2], "P3"

# P4 (12-16): x3 + x4 >=7
prob += x[3] + x[4] >= demandas[3], "P4"

# P5 (16-20): x4 + x5 >=10
prob += x[4] + x[5] >= demandas[4], "P5"

# P6 (20-24): x5 + x6 >=4
prob += x[5] + x[6] >= demandas[5], "P6"

# Resolver
prob.solve(pulp.PULP_CBC_CMD(msg=0))

# Resultados
print("Objetivo:", pulp.value(prob.objective))
for i in range(1,7):
    print(f"x{i}: {x[i].varValue}")

Objetivo: 26.0
x1: 0.0
x2: 10.0
x3: 0.0
x4: 12.0
x5: 0.0
x6: 4.0


La solución evita tomar turnos necesarios, entonces en los turnos 1, 3 y 5 no salen autobuses. Nótese que para satisfacer la demanda se empieza desde el turno 6, saliendo 4 autobuses, alcanzando hasta el turno 2, donde salen 10 autobuses y finalmente hasta el turno 4 que salen los 12 autobuses finales. 

#### Problema 5. 

Definimos las variables de decisión $x_{ij} \geq 0$ (millones de galones enviados diariamente de la refinería $i$ a la área de distribución $j$), donde $i=1,2,3$ (refinerías) y $j=1,2,3$ (áreas). No existe conexión de la refinería 1 a la área 3 ($x_{13}=0$ implícitamente).

**Datos clave:**  
- Capacidades (ofertas) $s_i$: 6, 5, 8 millones de galones (ref. 1, 2, 3).  
- Demandas $d_j$: 4, 8, 7 millones de galones (área 1, 2, 3).  
- Costo de transporte: $0.10 por 1000 galones por km, equivalente a $c_{ij} = 100 \times d_{ij}$ dólares por millón de galones, donde $d_{ij}$ son las distancias (km):  
  - Ref. 1: $d_{11}=120$, $d_{12}=180$, $d_{13}=-\infty$ (sin conexión).  
  - Ref. 2: $d_{21}=300$, $d_{22}=100$, $d_{23}=80$.  
  - Ref. 3: $d_{31}=200$, $d_{32}=250$, $d_{33}=120$.  

**Función objetivo** (minimizar costo total de transporte):  
$$
\min z = \sum_{i=1}^3 \sum_{j=1}^3 c_{ij} x_{ij}
$$  
con $c_{11}=12000$, $c_{12}=18000$, $c_{21}=30000$, $c_{22}=10000$, $c_{23}=8000$, $c_{31}=20000$, $c_{32}=25000$, $c_{33}=12000$ (y $c_{13}=\infty$ o $x_{13}=0$).

**Restricciones:**  
1. **Ofertas (capacidades de refinerías)**:  
$$
x_{11} + x_{12} = 6 \quad (\text{Ref. 1})
$$  
$$
x_{21} + x_{22} + x_{23} = 5 \quad (\text{Ref. 2})
$$  
$$
x_{31} + x_{32} + x_{33} = 8 \quad (\text{Ref. 3})
$$  

2. **Demandas (áreas de distribución)**:  
$$
x_{11} + x_{21} + x_{31} = 4 \quad (\text{Área 1})
$$  
$$
x_{12} + x_{22} + x_{32} = 8 \quad (\text{Área 2})
$$  
$$
x_{23} + x_{33} = 7 \quad (\text{Área 3, sin } x_{13})
$$  

3. **No negatividad**:  
$$
x_{ij} \geq 0, \quad \forall i,j \quad (x_{13}=0)
$$  

Este es un problema de programación lineal de transporte balanceado (oferta total=19 = demanda total=19) con 8 variables y 6 restricciones de igualdad.

In [9]:
import pulp

# Transportation Problem
prob = pulp.LpProblem("Transporte_Gasolina", pulp.LpMinimize)

# Variables: x_ij from refinery i=1,2,3 to area j=1,2,3 (millones de galones)
x11 = pulp.LpVariable("x11", lowBound=0)
x12 = pulp.LpVariable("x12", lowBound=0)
# x13 = 0 (no connection)

x21 = pulp.LpVariable("x21", lowBound=0)
x22 = pulp.LpVariable("x22", lowBound=0)
x23 = pulp.LpVariable("x23", lowBound=0)

x31 = pulp.LpVariable("x31", lowBound=0)
x32 = pulp.LpVariable("x32", lowBound=0)
x33 = pulp.LpVariable("x33", lowBound=0)

# Costos unitarios: 100 * distancia (en $ por millon galones)
c11 = 100 * 120  # 12000
c12 = 100 * 180  # 18000
# c13 infinite or 0 var

c21 = 100 * 300  # 30000
c22 = 100 * 100  # 10000
c23 = 100 * 80   # 8000

c31 = 100 * 200  # 20000
c32 = 100 * 250  # 25000
c33 = 100 * 120  # 12000

# Objetivo
prob += (c11*x11 + c12*x12 + c21*x21 + c22*x22 + c23*x23 + c31*x31 + c32*x32 + c33*x33)

# Capacidades (ofertas)
prob += x11 + x12 == 6, "Oferta1"
prob += x21 + x22 + x23 == 5, "Oferta2"
prob += x31 + x32 + x33 == 8, "Oferta3"

# Demandas
prob += x11 + x21 + x31 == 4, "Demanda1"
prob += x12 + x22 + x32 == 8, "Demanda2"
prob += x23 + x33 == 7, "Demanda3"  # No x13

# Resolver
prob.solve(pulp.PULP_CBC_CMD(msg=0))

# Resultados
print("Costo total:", pulp.value(prob.objective))
print("x11:", x11.varValue)
print("x12:", x12.varValue)
print("x21:", x21.varValue)
print("x22:", x22.varValue)
print("x23:", x23.varValue)
print("x31:", x31.varValue)
print("x32:", x32.varValue)
print("x33:", x33.varValue)

Costo total: 243000.0
x11: 4.0
x12: 2.0
x21: 0.0
x22: 5.0
x23: 0.0
x31: 0.0
x32: 1.0
x33: 7.0


La solución óptima fue enviar (en millones), de la refineria 1, 4 al área 1 y 2 al área2. Luego de la refinería 2, se mandará 5 galones al área 2 y por último de la refinería 3, se enviarán 1 galón al área 2 y 7 galones al área 3. Dando un costo total de 243,000

#### c)

**Datos clave modificados:**  
- Demandas: Área 1=4, Área 2=8, Área 3=4 millones de galones (total demanda=16; excedente total=3 millones).  
- Excedentes $e_i \geq 0$ (millones de galones no enviados por oleoducto):  
  - Ref. 1 y 2: Costo camión $1.50 y $2.20 por 100 galones (escalado: $15,000 y $22,000 por millón).  
  - Ref. 3: Costo interno = 0.  

**Función objetivo** (minimizar costo total, oleoductos + camiones):  
$$
\min z = \sum_{i=1}^3 \sum_{j=1}^3 c_{ij} x_{ij} + 15{,}000 e_1 + 22{,}000 e_2 + 0 \cdot e_3
$$  
(con $c_{ij}$ como en (a); $x_{13}=0$).

**Restricciones:**  
1. **Ofertas con excedentes**:  
$$
x_{11} + x_{12} + e_1 = 6 \quad (\text{Ref. 1})
$$  
$$
x_{21} + x_{22} + x_{23} + e_2 = 5 \quad (\text{Ref. 2})
$$  
$$
x_{31} + x_{32} + x_{33} + e_3 = 8 \quad (\text{Ref. 3})
$$  

2. **Demandas**:  
$$
x_{11} + x_{21} + x_{31} = 4 \quad (\text{Área 1})
$$  
$$
x_{12} + x_{22} + x_{32} = 8 \quad (\text{Área 2})
$$  
$$
x_{23} + x_{33} = 4 \quad (\text{Área 3})
$$  

3. **No negatividad**: $x_{ij}, e_i \geq 0 \quad (x_{13}=0)$.  

Este es un problema de programación lineal desbalanceado (oferta > demanda) con 11 variables y 6 restricciones.

In [10]:
import pulp

# Transportation Problem Modificado (demanda Area 3 = 4)
prob_mod = pulp.LpProblem("Transporte_Gasolina_Mod", pulp.LpMinimize)

# Variables: x_ij from refinery i=1,2,3 to area j=1,2,3 (millones de galones)
x11 = pulp.LpVariable("x11", lowBound=0)
x12 = pulp.LpVariable("x12", lowBound=0)
# x13 = 0 (no connection)

x21 = pulp.LpVariable("x21", lowBound=0)
x22 = pulp.LpVariable("x22", lowBound=0)
x23 = pulp.LpVariable("x23", lowBound=0)

x31 = pulp.LpVariable("x31", lowBound=0)
x32 = pulp.LpVariable("x32", lowBound=0)
x33 = pulp.LpVariable("x33", lowBound=0)

# Excedentes e_i
e1 = pulp.LpVariable("e1", lowBound=0)  # Ref 1 por camion
e2 = pulp.LpVariable("e2", lowBound=0)  # Ref 2 por camion
e3 = pulp.LpVariable("e3", lowBound=0)  # Ref 3 interno

# Costos unitarios pipelines: 0.10 por 1000 gal/km = 100 por km por millón gal
c11 = 100 * 120  # 12000
c12 = 100 * 180  # 18000

c21 = 100 * 300  # 30000
c22 = 100 * 100  # 10000
c23 = 100 * 80   # 8000

c31 = 100 * 200  # 20000
c32 = 100 * 250  # 25000
c33 = 100 * 120  # 12000

# Costos camiones: escalado por 100 gal a por millón (1e6 / 100 = 10000 unidades)
c_e1 = 1.50 * 10000  # 15000
c_e2 = 2.20 * 10000  # 22000
c_e3 = 0

# Objetivo
prob_mod += (c11*x11 + c12*x12 + c21*x21 + c22*x22 + c23*x23 + c31*x31 + c32*x32 + c33*x33 +
             c_e1*e1 + c_e2*e2 + c_e3*e3)

# Capacidades (ofertas con excedentes)
prob_mod += x11 + x12 + e1 == 6, "Oferta1"
prob_mod += x21 + x22 + x23 + e2 == 5, "Oferta2"
prob_mod += x31 + x32 + x33 + e3 == 8, "Oferta3"

# Demandas (Area 3 = 4)
prob_mod += x11 + x21 + x31 == 4, "Demanda1"
prob_mod += x12 + x22 + x32 == 8, "Demanda2"
prob_mod += x23 + x33 == 4, "Demanda3"  # Modificado

# Resolver
prob_mod.solve(pulp.PULP_CBC_CMD(msg=0))

# Resultados
print("Costo total:", pulp.value(prob_mod.objective))
print("x11:", x11.varValue)
print("x12:", x12.varValue)
print("x21:", x21.varValue)
print("x22:", x22.varValue)
print("x23:", x23.varValue)
print("x31:", x31.varValue)
print("x32:", x32.varValue)
print("x33:", x33.varValue)
print("e1:", e1.varValue)
print("e2:", e2.varValue)
print("e3:", e3.varValue)

Costo total: 207000.0
x11: 4.0
x12: 2.0
x21: 0.0
x22: 5.0
x23: 0.0
x31: 0.0
x32: 1.0
x33: 4.0
e1: 0.0
e2: 0.0
e3: 3.0


#### Problema 6

### 6. (Problema de asignación)

**costos** $c_{ij}$ (trabajador $i$ a tarea $j$):


**Variables**: $x_{ij} \in \{0,1\}$ (1 si asigna $i$ a $j$).

**Función objetivo** (minimizar costo total):  
$$
\min z = \sum_{i=1}^5 \sum_{j=1}^5 c_{ij} x_{ij}
$$

**Restricciones**:  
1. Cada trabajador una tarea: $\sum_{j=1}^5 x_{ij} = 1, \quad \forall i$.  
2. Cada tarea un trabajador: $\sum_{i=1}^5 x_{ij} = 1, \quad \forall j$.  


In [14]:
import pulp

# Matriz de costos (6x7)
costs = [
    [3, 8, 2, 10, 3, 3, 9],
    [2, 2, 7, 6, 5, 2, 7],
    [5, 6, 4, 5, 6, 6, 6],
    [4, 2, 7, 5, 9, 4, 7],
    [10, 3, 8, 4, 2, 3, 5],
    [3, 5, 4, 2, 3, 7, 8]
]

rows = len(costs)
cols = len(costs[0])

# Definimos el problema
prob = pulp.LpProblem("Asignacion", pulp.LpMinimize)

# Variables de decisión
x = pulp.LpVariable.dicts("x", (range(rows), range(cols)), cat="Binary")

# Función objetivo
prob += pulp.lpSum(costs[i][j] * x[i][j] for i in range(rows) for j in range(cols))

# Restricciones: cada trabajador (fila) debe tener 1 tarea
for i in range(rows):
    prob += pulp.lpSum(x[i][j] for j in range(cols)) == 1

# Restricciones: cada tarea (columna) como máximo a 1 trabajador
for j in range(cols):
    prob += pulp.lpSum(x[i][j] for i in range(rows)) <= 1

# Resolver
prob.solve()

# Resultados
print("Status:", pulp.LpStatus[prob.status])
print("Asignaciones óptimas:")

total_cost = 0
for i in range(rows):
    for j in range(cols):
        if pulp.value(x[i][j]) == 1:
            print(f"Trabajador {i+1} → Tarea {j+1} (costo = {costs[i][j]})")
            total_cost += costs[i][j]

print("Costo total mínimo =", total_cost)


Status: Optimal
Asignaciones óptimas:
Trabajador 1 → Tarea 1 (costo = 3)
Trabajador 2 → Tarea 6 (costo = 2)
Trabajador 3 → Tarea 3 (costo = 4)
Trabajador 4 → Tarea 2 (costo = 2)
Trabajador 5 → Tarea 5 (costo = 2)
Trabajador 6 → Tarea 4 (costo = 2)
Costo total mínimo = 15
